qiskit/opflow/list_ops/list_op.py

```diff
@@ -72,12 +72,19 @@ def __init__(self,
             identity - it accepts the list of values, and returns them in a list.
         """
         super().__init__()
-        self._oplist = list(oplist)
+        self._oplist = self._check_input_types(oplist)
         self._combo_fn = combo_fn
         self._coeff = coeff
         self._abelian = abelian
         self._grad_combo_fn = grad_combo_fn
 
+    def _check_input_types(self, oplist):
```

Comment:
Is this done with the aspect of supporting other types such as float as well as OperatorBase elsewhere that might get passed here. Normally type checks are only done if the action is different based on type, not to check the type is the sole type it expects. Especially since the method says it checks input types whereas its only checking type of oplist not all the 'inputs' ie params of this class

Reply:
* No, I have no intention to support `float` at the moment.
* `OperatorBase` is currently the only type it expects, but it is not enforced. In fact there is code that constructs `ListOp` and subclasses with elements, and we expect more code like it. This is meant to catch this error early and give a meaningful error message. Do not want to validate the content of `ListOp`, or to do it elsewhere?
* You mean `_check_input_types` is not an accurate name because it does not check type of all inputs to `__init__`? If so, I am not opposed to changing the name of this method.

Reply:
I guess I am questioning why validate the type of just one parameter and ignore the others, when you choose to do validation of types, which I was pointing out is normally not done. Maybe users get into more trouble with the oplist and, as you say, some weird error(s) that may later occur may make this tougher to figure than if something incorrect is passed to the other params. Hence its reasonable to do this validation for this specific param. 

As far as the name goes the function seems very specific to checking oplist (or at least some input that has to be a collection of OperatorBase instances) . Had it passed the param plus a list of valid types such that perhaps it was a more general checker the name would be more valid in my mind - it would have then checked the input against the given types.

Reply:
I'm not really opposed to checking the other arguments; it might be a good idea. But, the `oplist` is the problematic one. It seems the easiest place to inadvertently send the wrong thing. And in fact the code that does result in incorrect types in the oplist is internal. For example, the reduce method here returned a `float`, but the return value is expected to be of type `OperatorBase`.

I see that input types are not often validated. Maybe there is a good reason for this. But, there is a clear reason to validate it. We were discussing where the zeros came from in the bugs that led to this. Yes, you could track it down with a debugger or print statements or whatnot. But, after adding validation (but with the still-faulty `reduce` method) it is immediately obvious from the stack trace where there error is. I'm a bit puzzled why we put so much effort into the type annotations, if we don't really make an effort to stop bugs by validating types. In fact this is a good argument for single/multiple dispatch libraries, validation and mypy types are specified on one place.

Reply:
If this is the common case that fails and it helps users by validating this early so their problem is diagnosed early I have not real issue with this. The typehints document what you should be doing - like this is a one way street. They do not enforce it but non-compliance can be problematic! If we wanted type enforcement then we ought to consider a strongly typed language so the burden of type checking etc is not the guy doing the code.

Reply:
(missed your last comment yesterday) 

Looking at my comment above, I think my point may be buried a bit. I think you got it, but I'll repeat it for clarity. The documentation/annotation is maybe good enough for people deliberately calling the constructor from a high level. I added the check because there is construction that occurs in internal code using objects whose provenance is not 100% clear. So, the coders think they are probably sending the correct thing to the constructor, but in fact they are not. And the runtime error occurs somewhere down the line.

I agree that the person writing the code making adhoc checks is not a great state of affairs. There are quite a few such checks here and there in qiskit, and I don't enjoy seeing them.  But, I find they can be very effective in helping prevent bugs or identify them quickly. I suspect some were added after chasing down a bug that would have been quick to identify had the check been there in the first place.

This is one advantage of using single or multiple dispatch: A single annotation provides both documentation and type checking. At this point, I wouldn't think it should be used everywhere, but it could be used effectively.